File: aspnet-in-practice/08-09/CS/CoolMVBlog/Areas/Backoffice/Models/AuthorsService.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Web;
using CoolMvcBlog.Models;
using System.Web.Caching;
using CoolMvcBlog.Utils;

namespace CoolMvcBlog.Areas.Backoffice.Models
{
    public class AuthorsService
    {
        public static List<Author> GetAuthors()
        {
            HttpContext context = HttpContext.Current;
            if (context == null)
                throw new InvalidOperationException("A valid HttpContext is needed");

            var res = context.Cache["authors.all"] as List<Author>;
            if (res == null)
            {
                context.Cache.Insert("authors.all", 
                    res = ObjectContextModule.CurrentContext.AuthorSet.ToList(),
                    null, DateTime.Now.AddHours(1), Cache.NoSlidingExpiration);
            }

            return res;
        }
    }
}

File: aspnet-in-practice/08-09/VB/CoolMVCBlog/Areas/Backoffice/Models/AuthorsService.vb
Public Class AuthorsService
    Public Shared Function GetAuthors() As List(Of Author)
        Dim context = HttpContext.Current
        If context Is Nothing Then
            Throw New InvalidOperationException("A valid HttpContext is needed")
        End If

        Dim res = TryCast(context.Cache("authors.all"), List(Of Author))
        If res Is Nothing Then
            res = ObjectContextModule.CurrentContext.AuthorSet.ToList()
            context.Cache.Insert("authors.all", res, Nothing,
                                 DateTime.Now.AddHours(1), Cache.NoSlidingExpiration)
        End If

        Return res
    End Function
End Class
